pkg/engine/internal/executor/arrow_scalar_compare.go

```diff
@@ -136,3 +137,132 @@ func compareScalars(left, right scalar.Scalar, nullsFirst bool) (int, error) {
 
 	return 0, nil
 }
+
+// compareArrays compares values at the given indices from two arrays, returning:
+//
+// - -1 if left < right
+// - 0 if left == right
+// - 1 if left > right
+//
+// If nullsFirst is true, then null values are considered to sort before
+// non-null values.
+//
+// compareArrays returns an error if the two arrays are of different types,
+// or if the array type is not supported for comparison.
+func compareArrays(left, right arrow.Array, leftIdx, rightIdx int, nullsFirst bool) (int, error) {
+	leftNull := left == nil || !left.IsValid(leftIdx)
+	rightNull := right == nil || !right.IsValid(rightIdx)
+
+	// First, handle one or both of the values being null.
+	switch {
+	case leftNull && rightNull:
+		return 0, nil
+
+	case leftNull && !rightNull:
+		if nullsFirst {
+			return -1, nil
+		}
+		return 1, nil
+
+	case !leftNull && rightNull:
+		if nullsFirst {
+			return 1, nil
+		}
+		return -1, nil
+	}
+
+	if !arrow.TypeEqual(left.DataType(), right.DataType()) {
+		// We should never hit this, since compareRow is only called for two arrays
+		// coming from the same [arrow.Field].
+		return 0, errors.New("received arrays of different types")
+	}
+
+	// Fast-path: if both arrays reference the same underlying data and same index,
+	// they're equal. This is an optimization for common cases.
+	if left == right && leftIdx == rightIdx {
+		return 0, nil
+	}
+
+	// Switch on the array type to compare the values. This is only composed of
+	// types we know the query engine uses, and types that we know have clear
+	// sorting semantics.
+	//
+	// Unsupported types are treated as equal for consistent sorting, but
+	// otherwise it's up to the caller to detect unexpected sort types and reject
+	// the query.
+	switch left := left.(type) {
+	case *array.Binary:
+		right := right.(*array.Binary)
+		return bytes.Compare(left.Value(leftIdx), right.Value(rightIdx)), nil
+
+	case *array.Dictionary:
+		right := right.(*array.Dictionary)
+		leftDictIdx := left.GetValueIndex(leftIdx)
+		rightDictIdx := right.GetValueIndex(rightIdx)
+
+		// Get the dictionary arrays
+		leftDict := left.Dictionary()
+		rightDict := right.Dictionary()
+
+		// Compare the dictionary arrays at the respective indices
+		return compareArrays(leftDict, rightDict, leftDictIdx, rightDictIdx, nullsFirst)
+
+	case *array.Duration:
+		right := right.(*array.Duration)
+		return cmp.Compare(left.Value(leftIdx), right.Value(rightIdx)), nil
+
+	case *array.Float16:
+		right := right.(*array.Float16)
+		return left.Value(leftIdx).Cmp(right.Value(rightIdx)), nil
+
+	case *array.Float32:
+		right := right.(*array.Float32)
+		return cmp.Compare(left.Value(leftIdx), right.Value(rightIdx)), nil
```

Comment:
I don't think that we need to support all array types, only the the ones used by the engine.

Reply:
removed most of the non-relevant ones in https://github.com/grafana/loki/pull/19660/commits/7f5c3782e5fd99021b1674a63917cfeb97020f0f